include/data_sink.hpp

```diff
@@ -0,0 +1,900 @@
+#ifndef GNURADIO_DATA_SINK_HPP
+#define GNURADIO_DATA_SINK_HPP
+
+#include "circular_buffer.hpp"
+#include "dataset.hpp"
+#include "history_buffer.hpp"
+#include "node.hpp"
+#include "tag.hpp"
+
+#include <any>
+#include <chrono>
+
+namespace fair::graph {
+
+enum class blocking_mode { NonBlocking, Blocking };
+
+enum class trigger_observer_state {
+    Start,        ///< Start a new dataset
+    Stop,         ///< Finish dataset
+    StopAndStart, ///< Finish pending dataset, start a new one
+    Ignore        ///< Ignore tag
+};
+
+// TODO is the scope where want these?
+struct null_type {};
+
+// Until clang-format can handle concepts
+// clang-format off
+template<typename T>
+concept TriggerPredicate = requires(const T p, tag_t tag) {
+    { p(tag) } -> std::convertible_to<bool>;
+};
+
+template<typename T>
+concept TriggerObserver = requires(T o, tag_t tag) {
+    { o(tag) } -> std::convertible_to<trigger_observer_state>;
+};
+
+template<typename T>
+concept TriggerObserverFactory = requires(T f) {
+    { f() } -> TriggerObserver;
+};
+// clang-format on
+
+template<typename T>
+class data_sink;
+
+struct data_sink_query {
+    std::optional<std::string> sink_name;
+    std::optional<std::string> signal_name;
+
+    static data_sink_query
+    with_signal_name(std::string_view name) {
+        return { {}, std::string{ name } };
+    }
+
+    static data_sink_query
+    with_sink_name(std::string_view name) {
+        return { std::string{ name }, {} };
+    }
+};
+
+class data_sink_registry {
+    std::mutex            _mutex;
+    std::vector<std::any> _sinks;
+
+public:
+    // TODO this shouldn't be a singleton but associated with the flow graph (?)
```

Comment:
Yes, I concur. The registry should eventually be owned by the graph and injected to the nodes that need this. We may need to generalise this before doing so.... Maybe in the end simply iterating over a flattened graph-node hierarchy would be sufficient. :thinking: 

@wirew0rm @ivan-cukic what do you think?

Reply:
This should also be solved before we move the block to blocklib/, as the registry might have to live with the graph? Alternatively, the graph could provide an interface to query nodes, then the user could create the registry (should get another name then), which would query that interface with data_sink instances.

Reply:
For the time being this needs to be only documented. We'll make that move once the graph-scheduler interface advanced a bit more.

Reply:
Can you add to the TODO that mutexes/locking should be rethought once it becomes a part of the graph?

Reply:
Done